lang_qc/pacbio/qc_data.py

```diff
@@ -0,0 +1,106 @@
+# Foundation; either version 3 of the License, or (at your option) any later
+# version.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program. If not, see <http://www.gnu.org/licenses/>
+
+import re
+
+from ml_warehouse.schema import PacBioRunWellMetrics
+from pydantic import BaseModel, Field
+
+
+class QCDataWell(BaseModel):
+
+    smrt_link: dict = Field(title="URL components for a SMRT Link page")
+    binding_kit: dict = Field(default=None, title="Binding Kit")
+    control_num_reads: dict = Field(default=None, title="Number of Control Reads")
+    control_read_length_mean: dict = Field(
+        default=None, title="Control Read Length (bp)"
+    )
+    hifi_read_bases: dict = Field(default=None, title="CCS Yield (Gb)")
+    hifi_read_length_mean: dict = Field(default=None, title="CCS Mean Length (bp)")
+    local_base_rate: dict = Field(default=None, title="CCS Mean Length (bp)")
+    p0_num: dict = Field(default=None, title="P0 %")
+    p1_num: dict = Field(default=None, title="P1 %")
+    p2_num: dict = Field(default=None, title="P2 %")
+    polymerase_read_bases: dict = Field(default=None, title="Total Cell Yield (Gb)")
+    polymerase_read_length_mean: dict = Field(
+        default=None, title="Mean Polymerase Read Length (bp)"
+    )
+    movie_minutes: dict = Field(default=None, title="Run Time (hr)")
+
+    class Config:
+        orm_mode = True
+
+    @classmethod
+    def from_orm(cls, obj: PacBioRunWellMetrics):
+
+        # Introspect the class definition, get a dictionary of specs
+        # for properties with property names as the keys.
+        attrs = cls.schema()["properties"]
+
+        straight_map_attr_names = set(
+            [
+                "binding_kit",
+                "control_num_reads",
+                "control_read_length_mean",
+                "hifi_read_length_mean",
+                "local_base_rate",
+                "polymerase_read_length_mean",
+            ]
+        )
+
+        productive_zmws_num = 0
+        if obj.productive_zmws_num is not None:
+            productive_zmws_num = obj.productive_zmws_num
+
+        qc_data = {}
+
+        for name in attrs:
+
+            if name == "smrt_link":
+                qc_data[name] = {
+                    "run_uuid": obj.sl_run_uuid,
+                    "hostname": obj.sl_hostname,
+                }
+                continue
+
+            value = getattr(obj, name)
+
+            if (value is not None) and (name not in straight_map_attr_names):
+                if name == "movie_minutes":
+                    value = round(value / 60)
+                elif re.search("p(0|1|2)_num", name):
```

Comment:
or `elif name not in ["p0_num", "p1_num", "p2_num"]:` doesn't need an import and is probably faster.